--- csharp/Makefile.am.bak	2009-07-30 23:30:24.000000000 -0700
+++ csharp/Makefile.am	2009-07-30 23:35:48.000000000 -0700
@@ -5,74 +5,85 @@
 noinst_SCRIPTS = Google.ProtocolBuffers.dll
 
 Google.ProtocolBuffers.dll_sources = \
-	$(srcdir)/ProtocolBuffers/AbstractBuilder.cs			\
-	$(srcdir)/ProtocolBuffers/AbstractMessage.cs			\
-	$(srcdir)/ProtocolBuffers/ByteString.cs				\
-	$(srcdir)/ProtocolBuffers/CodedInputStream.cs			\
-	$(srcdir)/ProtocolBuffers/CodedOutputStream.cs			\
-	$(srcdir)/ProtocolBuffers/Collections/Dictionaries.cs		\
-	$(srcdir)/ProtocolBuffers/Collections/Lists.cs			\
-	$(srcdir)/ProtocolBuffers/Collections/ReadOnlyDictionary.cs	\
-	$(srcdir)/ProtocolBuffers/DescriptorProtos/DescriptorProtoFile.cs	\
-	$(srcdir)/ProtocolBuffers/DescriptorProtos/IDescriptorProto.cs	\
-	$(srcdir)/ProtocolBuffers/DescriptorProtos/PartialClasses.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/DescriptorBase.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/DescriptorPool.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/DescriptorUtil.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/DescriptorValidationException.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/EnumDescriptor.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/EnumValueDescriptor.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/FieldDescriptor.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/FieldMappingAttribute.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/FieldType.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/FileDescriptor.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/IDescriptor.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/IndexedDescriptorBase.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/MappedType.cs		\
-	$(srcdir)/ProtocolBuffers/Descriptors/MessageDescriptor.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/MethodDescriptor.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/PackageDescriptor.cs	\
-	$(srcdir)/ProtocolBuffers/Descriptors/ServiceDescriptor.cs	\
-	$(srcdir)/ProtocolBuffers/DynamicMessage.cs			\
-	$(srcdir)/ProtocolBuffers/ExtendableBuilder.cs			\
-	$(srcdir)/ProtocolBuffers/ExtendableMessage.cs			\
-	$(srcdir)/ProtocolBuffers/ExtensionInfo.cs			\
-	$(srcdir)/ProtocolBuffers/ExtensionRegistry.cs			\
-	$(srcdir)/ProtocolBuffers/FieldAccess/Delegates.cs		\
-	$(srcdir)/ProtocolBuffers/FieldAccess/ReflectionUtil.cs		\
-	$(srcdir)/ProtocolBuffers/FieldAccess/SingleEnumAccessor.cs	\
-	$(srcdir)/ProtocolBuffers/FieldAccess/SingleMessageAccessor.cs	\
-	$(srcdir)/ProtocolBuffers/FieldAccess/SinglePrimitiveAccessor.cs	\
-	$(srcdir)/ProtocolBuffers/FieldAccess/RepeatedPrimitiveAccessor.cs	\
-	$(srcdir)/ProtocolBuffers/FieldAccess/RepeatedEnumAccessor.cs	\
-	$(srcdir)/ProtocolBuffers/FieldAccess/IFieldAccessor.cs		\
-	$(srcdir)/ProtocolBuffers/FieldAccess/FieldAccessorTable.cs	\
-	$(srcdir)/ProtocolBuffers/FieldAccess/RepeatedMessageAccessor.cs	\
-	$(srcdir)/ProtocolBuffers/FieldSet.cs				\
-	$(srcdir)/ProtocolBuffers/GeneratedBuilder.cs			\
-	$(srcdir)/ProtocolBuffers/GeneratedSingleExtension.cs		\
-	$(srcdir)/ProtocolBuffers/GeneratedMessage.cs			\
-	$(srcdir)/ProtocolBuffers/GeneratedRepeatException.cs		\
-	$(srcdir)/ProtocolBuffers/IBuilder.cs				\
-	$(srcdir)/ProtocolBuffers/GeneratedExtensionBase.cs		\
-	$(srcdir)/ProtocolBuffers/IMessage.cs				\
-	$(srcdir)/ProtocolBuffers/InvalidProtocolBufferException.cs	\
-	$(srcdir)/ProtocolBuffers/IRpcChannel.cs			\
-	$(srcdir)/ProtocolBuffers/IRpcController.cs			\
-	$(srcdir)/ProtocolBuffers/IService.cs				\
-	$(srcdir)/ProtocolBuffers/Properties/AssemblyInfo.cs		\
-	$(srcdir)/ProtocolBuffers/RpcUtil.cs				\
-	$(srcdir)/ProtocolBuffers/TextFormat.cs				\
-	$(srcdir)/ProtocolBuffers/TextGenerator.cs			\
-	$(srcdir)/ProtocolBuffers/TextTokenizer.cs			\
-	$(srcdir)/ProtocolBuffers/UninitializedMessageException.cs	\
-	$(srcdir)/ProtocolBuffers/UnknownField.cs			\
-	$(srcdir)/ProtocolBuffers/UnknownFieldSet.cs			\
-	$(srcdir)/ProtocolBuffers/WireFormat.cs
+$(srcdir)/ProtoGen/DependencyResolutionException.cs  \
+$(srcdir)/ProtoGen/InvalidOptionsException.cs  \
+$(srcdir)/ProtoGen/GeneratorOptions.cs  \
+$(srcdir)/ProtoGen/Generator.cs  \
+$(srcdir)/ProtoGen/Program.cs  \
+$(srcdir)/ProtocolBuffers/IRpcController.cs  \
+$(srcdir)/ProtocolBuffers/GeneratedExtensionBase.cs  \
+$(srcdir)/ProtocolBuffers/Properties/AssemblyInfo.cs  \
+$(srcdir)/ProtocolBuffers/CodedOutputStream.cs  \
+$(srcdir)/ProtocolBuffers/ByteString.cs  \
+$(srcdir)/ProtocolBuffers/GeneratedSingleExtension.cs  \
+$(srcdir)/ProtocolBuffers/IBuilder.cs  \
+$(srcdir)/ProtocolBuffers/UnknownFieldSet.cs  \
+$(srcdir)/ProtocolBuffers/AbstractMessage.cs  \
+$(srcdir)/ProtocolBuffers/IRpcChannel.cs  \
+$(srcdir)/ProtocolBuffers/TextGenerator.cs  \
+$(srcdir)/ProtocolBuffers/TextTokenizer.cs  \
+$(srcdir)/ProtocolBuffers/DynamicMessage.cs  \
+$(srcdir)/ProtocolBuffers/UnknownField.cs  \
+$(srcdir)/ProtocolBuffers/GeneratedMessage.cs  \
+$(srcdir)/ProtocolBuffers/WireFormat.cs  \
+$(srcdir)/ProtocolBuffers/InvalidProtocolBufferException.cs  \
+$(srcdir)/ProtocolBuffers/DescriptorProtos/IDescriptorProto.cs  \
+$(srcdir)/ProtocolBuffers/DescriptorProtos/PartialClasses.cs  \
+$(srcdir)/ProtocolBuffers/DescriptorProtos/DescriptorProtoFile.cs  \
+$(srcdir)/ProtocolBuffers/GeneratedBuilder.cs  \
+$(srcdir)/ProtocolBuffers/TextFormat.cs  \
+$(srcdir)/ProtocolBuffers/ExtensionRegistry.cs  \
+$(srcdir)/ProtocolBuffers/ExtendableMessage.cs  \
+$(srcdir)/ProtocolBuffers/RpcUtil.cs  \
+$(srcdir)/ProtocolBuffers/FieldSet.cs  \
+$(srcdir)/ProtocolBuffers/UninitializedMessageException.cs  \
+$(srcdir)/ProtocolBuffers/AbstractBuilder.cs  \
+$(srcdir)/ProtocolBuffers/Collections/Lists.cs  \
+$(srcdir)/ProtocolBuffers/Collections/Dictionaries.cs  \
+$(srcdir)/ProtocolBuffers/Collections/PopsicleList.cs  \
+$(srcdir)/ProtocolBuffers/Collections/ReadOnlyDictionary.cs  \
+$(srcdir)/ProtocolBuffers/Delegates.cs  \
+$(srcdir)/ProtocolBuffers/ExtensionInfo.cs  \
+$(srcdir)/ProtocolBuffers/ExtendableBuilder.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/FieldDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/IndexedDescriptorBase.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/PackageDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/DescriptorUtil.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/MethodDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/FileDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/MappedType.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/ServiceDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/DescriptorValidationException.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/DescriptorBase.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/FieldMappingAttribute.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/EnumDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/IDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/DescriptorPool.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/FieldType.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/MessageDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/Descriptors/EnumValueDescriptor.cs  \
+$(srcdir)/ProtocolBuffers/MessageStreamIterator.cs  \
+$(srcdir)/ProtocolBuffers/CodedInputStream.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/IFieldAccessor.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/RepeatedMessageAccessor.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/SingleMessageAccessor.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/RepeatedPrimitiveAccessor.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/ReflectionUtil.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/RepeatedEnumAccessor.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/FieldAccessorTable.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/SinglePrimitiveAccessor.cs  \
+$(srcdir)/ProtocolBuffers/FieldAccess/SingleEnumAccessor.cs  \
+$(srcdir)/ProtocolBuffers/MessageStreamWriter.cs  \
+$(srcdir)/ProtocolBuffers/IMessage.cs  \
+$(srcdir)/ProtocolBuffers/IService.cs  \
+$(srcdir)/ProtocolBuffers/GeneratedRepeatExtension.cs 
+
+#$(srcdir)/ProtoGen/Properties/AssemblyInfo.cs
+
 
 Google.ProtocolBuffers.dll: $(Google.ProtocolBuffers.dll_sources)
 	cp $(srcdir)/ProtocolBuffers/Properties/Google.ProtocolBuffers.snk .
-	smcs -codepage:utf8 -debug -target:library -out:Google.ProtocolBuffers.dll \
+	../../installed-mono/bin/gmcs -codepage:utf8 -debug -target:library -out:Google.ProtocolBuffers.dll \
         $(Google.ProtocolBuffers.dll_sources) -keyfile:Google.ProtocolBuffers.snk -r:System.dll
 
 install-data-local: Google.ProtocolBuffers.dll
