Index: src/third_party/WebKit/JavaScriptCore/wtf/ThreadingPthreads.cpp
===================================================================
--- src/third_party/WebKit/JavaScriptCore/wtf/ThreadingPthreads.cpp	(revision 44837)
+++ src/third_party/WebKit/JavaScriptCore/wtf/ThreadingPthreads.cpp	(working copy)
@@ -225,11 +225,14 @@
 
 bool isMainThread()
 {
+	return true;
+/*
 #if PLATFORM(DARWIN) && !PLATFORM(CHROMIUM)
     return pthread_main_np();
 #else
     return currentThread() == mainThreadIdentifier;
 #endif
+*/
 }
 
 Mutex::Mutex()
