diff --git a/src/google/protobuf/compiler/csharp/csharp_enum_field.cc b/src/google/protobuf/compiler/csharp/csharp_enum_field.cc
index 00b6991..7debeea 100644
--- a/src/google/protobuf/compiler/csharp/csharp_enum_field.cc
+++ b/src/google/protobuf/compiler/csharp/csharp_enum_field.cc
@@ -206,7 +206,7 @@ GenerateBuilderMembers(io::Printer* printer) const {
     "  return this;\r\n"
     "}\r\n"
     "public Builder Clear$capitalized_name$() {\r\n"
-    "  result.$name$_ = pbc::Lists<$type$>.Empty;\r\n"
+    "  result.$name$_ = global::Google.ProtocolBuffers.Collections.Lists<$type$>.Empty;\r\n"
     "  return this;\r\n"
     "}\r\n");
 }
diff --git a/src/google/protobuf/compiler/csharp/csharp_extension.cc b/src/google/protobuf/compiler/csharp/csharp_extension.cc
index a9103e9..2e05bc9 100644
--- a/src/google/protobuf/compiler/csharp/csharp_extension.cc
+++ b/src/google/protobuf/compiler/csharp/csharp_extension.cc
@@ -58,11 +58,11 @@ void ExtensionGenerator::Generate(io::Printer* printer) {
     printer->Print(vars,
       "public static readonly\r\n"
       "  pb::GeneratedExtensionBase<scg::IList<$type$>> $name$ =\r\n"
-      "      pb::GeneratedRepeatExtension<$type$>.CreateInstance(Descriptor.Extensions[$index$]);\r\n");
+      "      global::Google.ProtocolBuffers.GeneratedRepeatExtension<$type$>.CreateInstance(Descriptor.Extensions[$index$]);\r\n");
   } else {
     printer->Print(vars,
       "public static readonly pb::GeneratedExtensionBase<$type$> $name$ =\r\n"
-      "      pb::GeneratedSingleExtension<$type$>.CreateInstance(Descriptor.Extensions[$index$]);\r\n");
+      "      global::Google.ProtocolBuffers.GeneratedSingleExtension<$type$>.CreateInstance(Descriptor.Extensions[$index$]);\r\n");
   }
 }
 
diff --git a/src/google/protobuf/compiler/csharp/csharp_file.cc b/src/google/protobuf/compiler/csharp/csharp_file.cc
index 4c0d744..597daf6 100644
--- a/src/google/protobuf/compiler/csharp/csharp_file.cc
+++ b/src/google/protobuf/compiler/csharp/csharp_file.cc
@@ -121,9 +121,9 @@ void FileGenerator::Generate(io::Printer* printer) {
   printer->Print(
     "public static pbd::FileDescriptor Descriptor {\r\n"
     "    get { return descriptor; }\r\n"
-    "}\r\n"
-    "private static readonly pbd::FileDescriptor descriptor = pbd::FileDescriptor.InternalBuildGeneratedFileFrom (\r\n"
-    "    new byte[] {");
+    "}\r\n\r\n"
+    "private static byte[] fileDescriptorBytes() {\r\n"
+    "    return new byte[] {");
   printer->Indent();
   printer->Indent();
   printer->Indent();
@@ -141,9 +141,23 @@ void FileGenerator::Generate(io::Printer* printer) {
     sprintf(buffer, "%02x", static_cast<uint8>(file_data[i]));
     printer->Print("0x$val$, ", "val", buffer);
   }
+
+  printer->Outdent();
   printer->Outdent();
+
+  printer->Print("\r\n};\r\n");
+
   printer->Outdent();
-  printer->Print("\r\n}, new pbd::FileDescriptor[] {\r\n");
+  printer->Outdent();
+
+  printer->Print(
+    "}\r\n"
+    "private static readonly pbd::FileDescriptor descriptor = pbd::FileDescriptor.InternalBuildGeneratedFileFrom (\r\n"
+    "    fileDescriptorBytes(),");
+  printer->Indent();
+  printer->Indent();
+
+  printer->Print("\r\n new global::Google.ProtocolBuffers.Descriptors.FileDescriptor[] {\r\n");
   for (int i = 0; i < file_->dependency_count(); i++) {
     printer->Print(
       "        $dependency$.Descriptor,\r\n",
diff --git a/src/google/protobuf/compiler/csharp/csharp_message.cc b/src/google/protobuf/compiler/csharp/csharp_message.cc
index 745cd4b..eddedbf 100644
--- a/src/google/protobuf/compiler/csharp/csharp_message.cc
+++ b/src/google/protobuf/compiler/csharp/csharp_message.cc
@@ -304,7 +304,7 @@ GenerateMessageSerializationMethods(io::Printer* printer) {
 
   if (descriptor_->extension_range_count() > 0) {
     printer->Print(
-      "pb::ExtendableMessage<$classname$, $classname$.Builder>.ExtensionWriter extensionWriter = CreateExtensionWriter(this);\r\n",
+      "global::Google.ProtocolBuffers.ExtendableMessage<$classname$, $classname$.Builder>.ExtensionWriter extensionWriter = CreateExtensionWriter(this);\r\n",
       "classname", descriptor_->name());
   }
 
diff --git a/src/google/protobuf/compiler/csharp/csharp_message_field.cc b/src/google/protobuf/compiler/csharp/csharp_message_field.cc
index 9c3ce49..4f3b33e 100644
--- a/src/google/protobuf/compiler/csharp/csharp_message_field.cc
+++ b/src/google/protobuf/compiler/csharp/csharp_message_field.cc
@@ -175,7 +175,7 @@ RepeatedMessageFieldGenerator::~RepeatedMessageFieldGenerator() {}
 void RepeatedMessageFieldGenerator::
 GenerateMembers(io::Printer* printer) const {
   printer->Print(variables_,
-    "private pbc::PopsicleList<$type$> $name$_ = new pbc::PopsicleList<$type$>();\r\n"
+    "private pbc::PopsicleList<$type$> $name$_ = new global::Google.ProtocolBuffers.Collections.Lists<$type$>.Empty;\r\n"
     "public scg::IList<$type$> $capitalized_name$List {\r\n"
     "  get { return $name$_; } \r\n"   // Will be unmodifiable by the time it's exposed
     "}\r\n"
diff --git a/src/google/protobuf/compiler/csharp/csharp_primitive_field.cc b/src/google/protobuf/compiler/csharp/csharp_primitive_field.cc
index 4bd6f49..69419f2 100644
--- a/src/google/protobuf/compiler/csharp/csharp_primitive_field.cc
+++ b/src/google/protobuf/compiler/csharp/csharp_primitive_field.cc
@@ -239,7 +239,7 @@ RepeatedPrimitiveFieldGenerator::~RepeatedPrimitiveFieldGenerator() {}
 void RepeatedPrimitiveFieldGenerator::
 GenerateMembers(io::Printer* printer) const {
   printer->Print(variables_,
-    "private pbc::PopsicleList<$type$> $name$_ = new pbc::PopsicleList<$type$>();\r\n"
+    "private pbc::PopsicleList<$type$> $name$_ = new global::Google.ProtocolBuffers.Collections.PopsicleList<$type$>();\r\n"
     "public scg::IList<$type$> $capitalized_name$List {\r\n"
     "  get { return $name$_; }\r\n"   // note:  unmodifiable list
     "}\r\n"
