===================================================================
--- src/video/cocoa/SDL_cocoawindow.m	(revision 4685)
+++ src/video/cocoa/SDL_cocoawindow.m	(working copy)
@@ -22,6 +22,11 @@
 #include "SDL_config.h"
 
 #include "SDL_syswm.h"
+struct SDL_XSysWMinfo
+{
+    SDL_version version;
+    int data;
+};
 #include "../SDL_sysvideo.h"
 #include "../../events/SDL_keyboard_c.h"
 #include "../../events/SDL_mouse_c.h"
@@ -601,12 +606,15 @@
 }
 
 SDL_bool
-Cocoa_GetWindowWMInfo(_THIS, SDL_Window * window, SDL_SysWMinfo * info)
+Cocoa_GetWindowWMInfo(_THIS, SDL_Window * window, SDL_SysWMinfo * ainfo)
 {
+    struct SDL_XSysWMinfo * info=(struct SDL_XSysWMinfo*)ainfo;
+
     NSWindow *nswindow = ((SDL_WindowData *) window->driverdata)->window;
 
     if (info->version.major <= SDL_MAJOR_VERSION) {
-        //info->window = nswindow;
+        NSView *view=[nswindow contentView];
+        info->data = (int)view;
         return SDL_TRUE;
     } else {
         SDL_SetError("Application not compiled with SDL %d.%d\n",
